pkg/server/event_handler.go

```diff
@@ -234,33 +241,88 @@ func (h *EventHandler[T, P]) HTTPHandler() http.Handler {
+	// 	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+	// 		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	// 	}
+	// 	return nil
+	// }
+
+	// fmt.Println("processErr: " + processErr.Error())
+
+	// If error is a retryable error, we return err
+	// so we can start retry.
+	if processErr != nil {
+		if !errors.Is(processErr, pmaperrors.ErrNonRetryable) {
+			return processErr
+		}
+		attr[nonRetryableError] = processErr.Error()
+		logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+		if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+			fmt.Println("Failure failed to send downstream")
+			return fmt.Errorf("failed to send failure event downstream: %w", err)
+		}
+		return nil
+	}
+	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	}
+	return nil
+	// if errors.Is(processErr, pmaperrors.ErrRetryable) {
+	// 	fmt.Println("THIS IS RETRYABLE ERROR")
+	// 	return fmt.Errorf("EventHandler encountered retryable error: %w", processErr)
+	// }
+
+	// If the error is not retryable, we use failerMessenger
+	// to send the message to corresponding topic.
+	// attr[nonRetryableError] = processErr.Error()
+	// logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+	// if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+	// 	fmt.Println("Failure failed to send downstream")
+	// 	return fmt.Errorf("failed to send failure event downstream: %w", err)
+	// }
+	// This is supposed to return nil,
+	// currently, all errors are not of type ErrRetryable
+	// but we handle them as retryable errors.
+	// We will change this to nil after we sort out
+	// how to differenate these errors.
+	// return nil
+}
+
+func (h *EventHandler[T, P]) generatePmapEventBytes(ctx context.Context, m pubsub.Message) ([]byte, error) {
 	// Get the GCS object as a proto message given GCS notification information.
 	p, err := h.getGCSObjectProto(ctx, m.Attributes)
 	if err != nil {
-		return fmt.Errorf("failed to get GCS object: %w", err)
+		return nil, fmt.Errorf("failed to get GCS object: %w", err)
 	}
 
-	// TODO(#20): we need to have a way to differentiate retryable err vs. not.
-	// For non-retryable error, we need to have them enter a different BQ table per design.
-	// Currently all error events are sent to downstream if failureMessenger is provided
-	// including those retried events.
 	var processErr error
 	for _, processor := range h.processors {
+		// TODO: we need to differentiate retryable vs none retryable error
+		// where processors might return.
 		if err := processor.Process(ctx, p); err != nil {
-			processErr = fmt.Errorf("failed to process object: %w", err)
+			processErr = fmt.Errorf("processor error: %w: failed to process object: %w", pmaperrors.ErrNonRetryable, err)
```

Comment:
If you change the definition of the retryable and nontryable error of pmap error pkg, then you don't have to construct the error in a such complex way.